certora/specs/DoubleEndedQueue.spec

```
@@ -0,0 +1,240 @@
+}
+
+// Could be broken in theory, but not in practice
+function endLimited(int128 end) returns bool {
+    return max_int128() > to_mathint(end); // not inclusive so it can pushBack
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: end is larger or equal than begin if values are safely assumed not to overflow.                          │
+| NOTE: This assumption make the proofs unsound, but they can't realistically overflow                                |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant boundariesConsistency()
+    end() >= begin()
+    {
+        preserved {
+            require beginLimited(begin());
+            require endLimited(end());
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: length is end minus begin                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant lengthConsistency()
+    length() == absolute_diff(to_mathint(end()), to_mathint(begin()))
+    {
+        preserved {
+            requireInvariant boundariesConsistency;
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: out of bounds is always 0                                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant outOfBoundsDefault(int128 key)
+    (
+        key < begin() &&  // Is before beginning
+        key >= end() // Is after ending
+    ) => at_(to_uint256(key)) == 0
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: empty() is length 0 and no element exists                                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant emptiness(int128 key)
+    empty() <=> length() == 0 && at_(to_uint256(key)) == 0
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: front() and back() point to the [begin, end) range limits                                                │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant pointerCorrectness()
+    back() == at_(to_uint256(begin())) && front() == at_(to_uint256(end()) - 1)
```

You are missing the logic for `at`

```suggestion
    back() == at_(0) && front() == at_(length() - 1)
```